File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoCore/PressureConfig.swift
import Foundation

/// Validated configuration for a `PressureEngine`.
public struct PressureConfig: Sendable, Equatable {
    public var tokenCoefficient: Double
    public var toolCoefficient: Double
    public var depthCoefficient: Double
    public var dissipationPerStep: Double
    public var dissipationPerSecond: Double
    public var escalationThreshold: Double
    public var releaseThreshold: Double
    public var postReleaseLock: Bool

    public init(
        tokenCoefficient: Double = 0.015,
        toolCoefficient: Double = 0.08,
        depthCoefficient: Double = 0.05,
        dissipationPerStep: Double = 0.02,
        dissipationPerSecond: Double = 0.0,
        escalationThreshold: Double = 0.85,
        releaseThreshold: Double = 0.95,
        postReleaseLock: Bool = true
    ) {
        self.tokenCoefficient = tokenCoefficient
        self.toolCoefficient = toolCoefficient
        self.depthCoefficient = depthCoefficient
        self.dissipationPerStep = dissipationPerStep
        self.dissipationPerSecond = dissipationPerSecond
        self.escalationThreshold = escalationThreshold
        self.releaseThreshold = releaseThreshold
        self.postReleaseLock = postReleaseLock
    }

    public static let defaults = PressureConfig()

    /// Throws `ConfigError` if any field is out of range.
    public func validate() throws {
        let nonNeg: [(String, Double)] = [
            ("token_coefficient", tokenCoefficient),
            ("tool_coefficient", toolCoefficient),
            ("depth_coefficient", depthCoefficient),
            ("dissipation_per_step", dissipationPerStep),
            ("dissipation_per_second", dissipationPerSecond),
        ]
        for (name, val) in nonNeg {
            if val < 0 {
                throw ConfigError("\(name) must be non-negative (got \(val))")
            }
        }
        if escalationThreshold < 0 || escalationThreshold > 1 {
            throw ConfigError(
                "escalation_threshold must be in [0, 1] (got \(escalationThreshold))")
        }
        if releaseThreshold < 0 || releaseThreshold > 1 {
            throw ConfigError(
                "release_threshold must be in [0, 1] (got \(releaseThreshold))")
        }
        if releaseThreshold <= escalationThreshold {
            throw ConfigError(
                "release_threshold must exceed escalation_threshold (\(releaseThreshold) <= \(escalationThreshold))")
        }
    }
}

File: IAIso-v5.0/core/iaiso-php/src/Core/PressureConfig.php
<?php

declare(strict_types=1);

namespace IAIso\Core;

/** Validated configuration for a {@see PressureEngine}. */
final class PressureConfig
{
    public function __construct(
        public readonly float $tokenCoefficient = 0.015,
        public readonly float $toolCoefficient = 0.08,
        public readonly float $depthCoefficient = 0.05,
        public readonly float $dissipationPerStep = 0.02,
        public readonly float $dissipationPerSecond = 0.0,
        public readonly float $escalationThreshold = 0.85,
        public readonly float $releaseThreshold = 0.95,
        public readonly bool $postReleaseLock = true,
    ) {
    }

    public static function defaults(): self
    {
        return new self();
    }

    public static function builder(): PressureConfigBuilder
    {
        return new PressureConfigBuilder();
    }

    /** Throws {@see ConfigException} if any field is out of range. */
    public function validate(): void
    {
        $nonNeg = [
            'token_coefficient'      => $this->tokenCoefficient,
            'tool_coefficient'       => $this->toolCoefficient,
            'depth_coefficient'      => $this->depthCoefficient,
            'dissipation_per_step'   => $this->dissipationPerStep,
            'dissipation_per_second' => $this->dissipationPerSecond,
        ];
        foreach ($nonNeg as $name => $val) {
            if ($val < 0) {
                throw new ConfigException("$name must be non-negative (got $val)");
            }
        }
        if ($this->escalationThreshold < 0.0 || $this->escalationThreshold > 1.0) {
            throw new ConfigException(
                "escalation_threshold must be in [0, 1] (got {$this->escalationThreshold})");
        }
        if ($this->releaseThreshold < 0.0 || $this->releaseThreshold > 1.0) {
            throw new ConfigException(
                "release_threshold must be in [0, 1] (got {$this->releaseThreshold})");
        }
        if ($this->releaseThreshold <= $this->escalationThreshold) {
            throw new ConfigException(
                "release_threshold must exceed escalation_threshold ({$this->releaseThreshold} <= {$this->escalationThreshold})");
        }
    }
}
